common/benchmark_utils.h

```diff
@@ -0,0 +1,355 @@
+ * @var hdr_allocd   keeps track of whether the hdr was allocd internally.
+ * @var hdr_printed  keeps track of whether the hdr has been printed to
+ * the console yet.
+ * @var fname_exists whether the file exists or not.
+ */
+static struct info_t {
+	char *fname;
+	struct stat stats;
+	FILE *fp;
+
+	char *hdr;
+	size_t hdr_len;
+	size_t cols;
+	size_t date_col;
+
+	char *date_str;
+	int date_len;
+
+	int hdr_allocd;
+	int hdr_printed;
+	int fname_exists;
+} info;
+
+/* Prototypes. */
+static inline int init_info(char *hdr, char *path);
+static inline void fini_info();
+static inline void print_data(double data);
+
+/*******************************************************************************
+ * Internal helper functions
+ ******************************************************************************/
+static inline void __set_cols()
+{
+	info.cols = 0;
+
+	if (info.hdr) {
+		char *tmp = strdup(info.hdr), *prev;
+		while (tmp) {
+			prev = tmp;
+			tmp = strchr(tmp + 1, ',');
+
+			if (tmp)
+				*tmp = 0;
+
+			if (!strcasecmp("date", prev)) {
+				info.date_col = info.cols;
+				info.date_len = strlen(info.date_str);
+			}
+
+			info.cols++;
+		}
+		free(tmp);
+	}
+}
+
+/**
+ * Get the hdr from an existing csv file specified by info.fname.
+ */
+static inline void __pull_and_set_hdr()
+{
+	char buf[4096] = {0};
+	int cnt = 1;
+
+	if (!info.hdr) {
+		info.hdr_len = 4096;
+		info.hdr = malloc(info.hdr_len);
+		assert(info.hdr);
+	} else {
+		fprintf(stderr, "hdr is already set.\n");
+		return;
+	}
+
+	while(fgets(buf, 4096, info.fp)) {
+		if (strrchr(buf, '\n')) { /* end of first line */
+			memcpy(info.hdr + (cnt - 1) * 4096, buf, strlen(buf));
+			break;
+		} else {
+			info.hdr_len = ++cnt * 4096;
+			info.hdr = realloc(info.hdr, info.hdr_len);
+			memcpy(info.hdr + (cnt - 1) * 4096, buf, 4096);
+		}
+	}
+
+	__set_cols();
+}
+
+#if 0
+/**
+ * Print the data to the given file specified by info.fname.
+ */
+static inline void __print_table(double data)
+{
+	static unsigned int col;
+
+	if (!info.hdr_printed) {
+		char *tmp = strdup(info.hdr), *ptr, *cur;
+
+
+		int i = 0;
+
+		for (cur = tmp; i < info.cols - 1; i++) {
+			ptr = strchr(cur, ',');
+
+			if (ptr)
+				*ptr = 0;
+
+			fprintf(stdout, "%-*s", FIELD_WIDTH, cur);
+
+			cur += strlen(cur) + 1;
+		}
+
+		fprintf(stdout, "%s.\n", cur);
+
+		free(tmp);
+		info.hdr_printed = 1;
+	}
+
+	if (info.date_len && col == info.date_col) {
+		time_t t;
+		struct tm *lt;
+
+		time(&t);
+		lt = localtime(&t);
+
+		if (lt) {
+			fprintf(stdout, info.date_str, lt->tm_year,
+				lt->tm_mon, lt->tm_mday, lt->tm_hour, lt->tm_min,
+				lt->tm_sec);
+			fprintf(stdout, ".");
+		}
+		col = (col + 1) % info.cols;
+	}
+
+	fprintf(stdout, col != (info.cols - 1) ? "%*.*lf" : "%*.*lf\n",
+		FIELD_WIDTH, FLOAT_PRECISION, data);
+
+	fflush(stdout);
+
+	col = (col + 1) % info.cols;
+}
+#endif
+
+/**
+ * Print the data to the given file specified by info.fname.
+ */
+static inline void __print_csv(double data)
+{
```

Comment:
> Is it always assumed that the data will be a double?

Yes, we could have a void ptr passed in and require that the user write their own casting functions or pass in a conversion character string but I decided keeping the data as a double to keep things simple.